engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/TypeSignatures.scala

```diff
@@ -169,6 +176,36 @@ case object TypeSignatures extends IRPass {
     }
   }
 
+  private def resolveDefinitionData(
+    data: IR.Module.Scope.Definition.Data
+  ): IR.Module.Scope.Definition.Data = {
+    data.copy(
```

Comment:
We should eventually start using a smarter `copy` that only copies if `data.arguments.map(resolvedArgument) != data.arguments`. But that optimization can be done in multiple places so can be done later.

Reply:
Am I right that `transformExpressions` does such kind of optimization a bit? My understanding is that `transformExpression` callback may specify partial `match` on just a few `IR.Expression.SubTypes` and only those get duplicated - the rest of the `IR` tree remains unchanged.

Shall we prefer the use of `transformExpressions` over `copy`? Are there any places to perform such a replacement in current code? 

 > start using a smarter copy
 
 Is there such a "smarter copy" already or do we have to write one? Can we create such "no-copy-if-unmodified" `copy` in a generic way or do we have to manually override each `copy` implementation or wrap each `copy` invocation?